### PR TITLE
Deimos gain/RN

### PR DESCRIPTION
**Pull request by debora-pe**
Now the gain and readnoise for DEIMOS are taken from a series of measurements performed by WMKO between 2008-2022. The measurements that are closest in time to the observations will be used.

### CODE REVIEW

**Review by bpholden — APPROVED** [review]
Do we want to have it fail to a particular value if the HDU is munged? Like the current values or most recently measured values in the file?

File: pypeit/data/utils.py [hunk 1]
```diff
@@ -225,6 +229,51 @@ def get_skisim_filepath(skisim_file):
     # Return the path to the `skisim` file
     return skisim_path
 
+def get_spectrograph_gain_ronoise_dirpath(spectrograph_name):
```
Comment (tbowers7) [inline pypeit/data/utils.py]: Because the files in `pypeit/data/spectrograph` are not overly large, we don't need to resort to only keeping them on GitHub.  As such, this new method is not needed.

File: pypeit/data/utils.py [hunk 2]
```diff
@@ -23,8 +23,8 @@
 
 __all__ = ['Paths', 'load_telluric_grid', 'load_thar_spec',
            'load_sky_spectrum', 'get_reid_arxiv_filepath',
-           'get_skisim_filepath', 'get_sensfunc_filepath',
-           'fetch_remote_file', 'write_file_to_cache']
+           'get_skisim_filepath', 'get_spectrograph_gain_ronoise_dirpath',
```
Comment (tbowers7) [inline pypeit/data/utils.py]: See comment attached to function, but this is not needed for your use case.

File: pypeit/data/utils.py [hunk 3]
```diff
@@ -70,6 +70,7 @@ def __init__(cls, *args, **kwargs):
         # Other
         cls._sky_spec = os.path.join(cls._data, 'sky_spec')
         cls._static_calibs = os.path.join(cls._data, 'static_calibs')
+        cls._spectrographs = os.path.join(cls._data, 'spectrographs')
```
Comment (tbowers7) [inline pypeit/data/utils.py:133]: Yes, this part is needed.

File: pypeit/data/utils.py [hunk 4]
```diff
@@ -124,6 +125,9 @@ def sky_spec(cls):
     @property
     def static_calibs(cls):
         return check_isdir(cls._static_calibs)
+    @property
+    def spectrographs(cls):
+        return check_isdir(cls._spectrographs)
```
Comment (tbowers7) [inline pypeit/data/utils.py:190]: Yes, this is needed -- sorry it didn't make it into the previous version.

File: pypeit/spectrographs/keck_deimos.py [hunk 5]
```diff
@@ -204,10 +206,56 @@ def get_detector_par(self, det, hdu=None):
...
+            darkcurr=3.70,
             gain=np.atleast_1d(1.230),
             ronoise=np.atleast_1d(2.580),
         ))
+
+        if hdu is not None:
+            # raw frame date in mjd
+            date = time.Time(self.get_meta_value(self.get_headarr(hdu), 'mjd'), format='mjd').value
+            # get the measurements files
+            measure_files = np.array(glob.glob(f'{data.get_spectrograph_gain_ronoise_dirpath("keck_deimos")}/*'))
```
Comment (tbowers7) [inline pypeit/spectrographs/keck_deimos.py]: This would change to:

`measure_files = np.array(glob.glob(os.path.join(data.Paths.spectrographs, "keck_deimos", "*")))`

**Review by tbowers7 — CHANGES_REQUESTED** [review]
Thanks for working to add the `spectrographs` data to the new framework for package data!

Basically the problem you were running into is that the function `fetch_remote_file()` is looking for files in the `develop` or version-tagged branch of the repo -- if you are adding new files, they don't exist there yet until the PR is merged.  (`TODO:` I need to add documentation to `pypeit/data/utils.py` specifying that if new files are added to the remote-fetch scheme, they need to be added to `develop` via a separate PR _before_ the PR that adds a new `get_*_filepath` method.)

In short, because the files in `pypeit/data/spectrographs` are generally small, we don't need to resort to the remote-fetch for them.  All that is needed is your addition of the `spectrographs` attribute to the `Paths` class, and pointing the load routine in `keck_deimos.py` to that path.  You should be able to add your files to the `pypeit/data/spectrographs` directory with this PR as you normally would have prior to my changes.

**Comment by kbwestfall** [comment]
There are already two reviewers on this, but I thought I would ask, did you forget to add the files in `pypeit/data/spectrographs/keck_deimos/` ?  I don't see them in the PR.

File: pypeit/data/utils.py (hunk 1 above)
Comment (deborape) [reply pypeit/data/utils.py]: > Do we want to have it fail to a particular value if the HDU is munged? Like the current values or most recently measured values in the file?
@bpholden if the hdu is not available, we'll use the value that we were using before. For the `MJD` header keyword, we have already a system in place for which if `MJD` does not exist, we'll use the observation date and convert it to mjd. I think that should secure us that this code will not crash.

**Comment by debora-pe** [comment]
> There are already two reviewers on this, but I thought I would ask, did you forget to add the files in `pypeit/data/spectrographs/keck_deimos/` ? I don't see them in the PR.

Thanks @kbwestfall. Yes, I was having problem updating them. I just realized that the files are `.log` which are "gitignored". So, I now added and exception  in `.gitignore`

**Comment by debora-pe** [comment]
> Do we want to have it fail to a particular value if the HDU is munged? Like the current values or most recently measured values in the file?
@bpholden if the hdu is not available, we'll use the value that we were using before. For the `MJD` header keyword, we have already a system in place for which if `MJD` does not exist, we'll use the observation date and convert it to mjd. I think that should secure us that this code will not crash.

File: pypeit/data/utils.py (hunk 1 above)
Comment (debora-pe) [reply pypeit/data/utils.py]: Removed!

File: pypeit/data/utils.py (hunk 2 above)
Comment (debora-pe) [reply pypeit/data/utils.py]: Removed

File: pypeit/spectrographs/keck_deimos.py (hunk 5 above)
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py]: Great thanks!

**Review by tbowers7 — APPROVED** [review]

**Comment by tbowers7** [comment]
Looks good @deborape !  Thanks!

**Comment by debora-pe** [comment]
Run Dev Suite on DEIMOS data:

```
 ./pypeit_test develop -i keck_deimos -t 8

Running tests on the following instruments:
    keck_deimos

Reducing data from keck_deimos for the following setups:
    600ZD_M_6500
    600ZD_tilted
    1200G_M_7750
    830G_LVM_8400
    830G_M_8100_26
    830G_M_8500
    830G_L_8100
    1200B_M_5200
    1200G_M_5500
    900ZD_M_6000
    1200B_LVM_5200
    900ZD_LVM_5500
    830G_M_9000_dither
    QL

Running tests in 8 parallel processes
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/600ZD_M_6500 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_8500 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/1200B_M_5200 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/QL pypeit_ql
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/900ZD_M_6000 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/1200G_M_7750 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_9000_dither pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/600ZD_tilted pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/QL pypeit_ql
 1 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_L_8100 pypeit (without masters)
 2 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_8500 pypeit (without masters)
 2 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_8500 pypeit_multislit_flexure
 3 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_8500 pypeit_multislit_flexure
 3 passed/ 0 failed/ 0 skipped STARTED keck_deimos/1200G_M_5500 pypeit (without masters)
 4 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/900ZD_M_6000 pypeit (without masters)
 4 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_LVM_8400 pypeit (without masters)
 5 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/1200B_M_5200 pypeit (without masters)
 5 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_8100_26 pypeit (without masters)
 6 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/600ZD_M_6500 pypeit (without masters)
 6 passed/ 0 failed/ 0 skipped STARTED keck_deimos/600ZD_M_6500 pypeit_flux
 7 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/600ZD_M_6500 pypeit_flux
 7 passed/ 0 failed/ 0 skipped STARTED keck_deimos/900ZD_LVM_5500 pypeit (without masters)
 8 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/1200G_M_5500 pypeit (without masters)
 8 passed/ 0 failed/ 0 skipped STARTED keck_deimos/1200B_LVM_5200 pypeit (without masters)
 9 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_L_8100 pypeit (without masters)
10 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/600ZD_tilted pypeit (without masters)
11 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_8100_26 pypeit (without masters)
12 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_LVM_8400 pypeit (without masters)
13 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/900ZD_LVM_5500 pypeit (without masters)
13 passed/ 0 failed/ 0 skipped STARTED keck_deimos/900ZD_LVM_5500 pypeit_sensfunc
14 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/900ZD_LVM_5500 pypeit_sensfunc
14 passed/ 0 failed/ 0 skipped STARTED keck_deimos/900ZD_LVM_5500 pypeit_flux
15 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/900ZD_LVM_5500 pypeit_flux
16 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/1200G_M_7750 pypeit (without masters)
17 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_9000_dither pypeit (without masters)
17 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_9000_dither pypeit_coadd_2dspec
18 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/1200B_LVM_5200 pypeit (without masters)
19 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_9000_dither pypeit_coadd_2dspec

--- PYPEIT DEVELOPMENT SUITE PASSED 19/19 TESTS (Masters ignored) ---
Testing Started at 2022-04-29T12:52:11.402647
Testing Completed at 2022-04-29T14:40:06.430874
Total Time: 1:47:55.028227
```

**Comment by debora-pe** [comment]
merging.
